sdk/src/metrics/export/periodic_exporting_metric_reader.cc

```diff
@@ -191,7 +193,11 @@ bool PeriodicExportingMetricReader::OnShutDown(std::chrono::microseconds timeout
 {
   if (worker_thread_.joinable())
   {
-    cv_.notify_one();
+    {
+      //ensure that `cv_` is awaiting, and the update doesn't get lost
+      std::unique_lock<std::mutex> lk(cv_m_);
```

Comment:
Do we need lock `cv_m_` here? Should we notify background thread immediately?

Reply:
Yes, this is needed to ensure that the CV is [being waited on](https://github.com/open-telemetry/opentelemetry-cpp/blob/2b357fbe027afb0c65a6ec8c938841a0f75e97f4/sdk/src/metrics/export/periodic_exporting_metric_reader.cc#L64) otherwise the notify gets lost.

Reply:
I was thinking about a lockless way to solve this problem. Or the critical section may be a little long to make more conflict.

Reply:
If it is possible for multiple threads to call flush, should the condition broadcast with notify_all() instead ?

Reply:
> @owent
> 
> About related issues like #2574
> 
> If it is possible for multiple threads to call flush/shutdown, should the condition broadcast with notify_all() instead ?

Both `notify_all` and `notify_one` may be missed when `wait_for` is called between predicate function and the internal `lock` function are called. I was thinking to use a atomic int64 to record the force flush sequence and wait only when notified sequence is less than flush sequence. We may assume int64 will not overflow in a single process.

Reply:
@owent do you mean tag each incoming request with a int64, and on a flush request use the current counter value to wait for all `before` requests to complete? 
The notify_all makes sense since we can have multiple waiters, and i can add that.

Reply:
Yes , `notify_all` would be great. #2584 may also solve the problem which this PR fixes.